### PR TITLE
Classification models

### PR DESCRIPTION
**Pull request by amadeuspzs**
# Summary

**What/Why**

Adds initial classification models, in draft form.

**Your approach**

New notebook with gridsearchcv optimised classification models.

**What to review**

06a Modelling - classification.ipynb

# Checklists

This pull/merge request meets the following requirements:

- [X] code runs
- [X] [developments are ethical][data-ethics-framework]
- [X] developments are secure (no data or notebook output)
- [X] you have made proportionate checks that the code works correctly
- [X] [minimum usable documentation][agilemodeling] written in the `docs` folder

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: notebooks/06a Modelling - classification.ipynb [hunk 1]
```diff
@@ -0,0 +1,835 @@
...
+   "outputs": [],
+   "source": [
+    "model_name = \"decisiontree\"\n",
+    "\n",
+    "# define gridsearch parameters\n",
+    "gsc = GridSearchCV(\n",
+    "    estimator=DecisionTreeClassifier(class_weight=weights.to_dict()),\n",
+    "    param_grid={\"max_depth\": [5, 10, None]},\n",
+    "    cv=5,\n",
+    "    scoring=\"accuracy\",\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:375]: Should this be accuracy? You've used class weights but the dataset is still imbalanced so I'd expect this metric to be deceptively high.

File: notebooks/06a Modelling - classification.ipynb [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+    "        \"max_depth\": [5, 10, None],\n",
+    "    },\n",
+    "    cv=5,\n",
+    "    scoring=\"balanced_accuracy\",\n",
+    "    verbose=1,\n",
+    "    n_jobs=-1,\n",
+    "    refit=True,\n",
+    ")\n",
+    "\n",
+    "# takes ~1 mins to run on a STANDARD_DS3_V2\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:517]: Minor point but are these times accurate? I trained the XGBoost model and it took ~8 mins. Just checking that something isn't wrong in my setup that leads to the longer train time.

File: notebooks/06a Modelling - classification.ipynb [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "model_name = \"xgboost\"\n",
+    "\n",
+    "# define gridsearch parameters\n",
+    "gsc = GridSearchCV(\n",
+    "    estimator=XGBClassifier(weight=weights.to_dict(), random_state=42),\n",
+    "    param_grid={\n",
+    "        \"n_estimators\": [1, 5],\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:506]: This range of estimators looks low to me? SKLearn default is 100 and I'd expect you'd want to try higher than this too?

File: notebooks/06a Modelling - classification.ipynb [hunk 4]
```diff
@@ -0,0 +1,835 @@
...
+    "        \"learning_rate\": [0.01, 0.1, 1],\n",
+    "        \"max_depth\": [5, 10, None],\n",
+    "    },\n",
+    "    cv=5,\n",
+    "    scoring=\"balanced_accuracy\",\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:511]: Does using balanced accuracy increase the chance of False Positives as it focuses on Recall?

File: notebooks/06a Modelling - classification.ipynb (hunk 1 above)
Comment (amadeuspzs) [reply notebooks/06a Modelling - classification.ipynb:375]: Excellent catch. Will update

File: notebooks/06a Modelling - classification.ipynb (hunk 2 above)
Comment (amadeuspzs) [reply notebooks/06a Modelling - classification.ipynb:517]: I have found some variability on Azure ML when training; is ~ order of magnitude friendly? If not I can clarify

File: notebooks/06a Modelling - classification.ipynb (hunk 3 above)
Comment (amadeuspzs) [reply notebooks/06a Modelling - classification.ipynb:506]: Great catch. From https://xgboost.readthedocs.io/en/stable/python/examples/sklearn_parallel.html?highlight=n_estimators#demo-for-using-xgboost-with-sklearn I'll use [50, 100, 200]

File: notebooks/06a Modelling - classification.ipynb [hunk 5]
```diff
@@ -0,0 +1,835 @@
...
+    "# Modeling\n",
+    "\n",
+    "Strategy is to try a number of classification models with:\n",
+    "\n",
+    "* GridsearchCV for hyperparameter tuning with cross validation, refitting full training set for best model\n",
+    "* Test all final models against the held-out test set.\n",
+    "* Explore feature importance of best performing model\n",
+    "* Explore fairness (next notebook) of best performing model\n",
+    "\n",
+    "Logistic regression models are excluded due to significant multi-collinearity (see analysis in Regression notebook)."
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:302]: I think we should consider LogReg using L1/L2 regularisation to try and alleviate multicollinearity

File: notebooks/06a Modelling - classification.ipynb [hunk 6]
```diff
@@ -0,0 +1,865 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1653046713351
+    }
+   },
+   "outputs": [],
+   "source": [
+    "weights = X_train.risk.value_counts().min() / X_train.risk.value_counts().sort_index()\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:282]: Up to you but the models you use all have a class_weight argument where 'balanced' is an option which will do this for you.

File: notebooks/06a Modelling - classification.ipynb [hunk 7]
```diff
@@ -0,0 +1,865 @@
...
+    }
+   },
+   "outputs": [],
+   "source": [
+    "model_name = \"decisiontree\"\n",
+    "\n",
+    "# define gridsearch parameters\n",
+    "gsc = GridSearchCV(\n",
+    "    estimator=DecisionTreeClassifier(class_weight=weights.to_dict()),\n",
+    "    param_grid={\"max_depth\": [5, 10, None]},  # default None\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:380]: Be aware that no max depth for Decision Tree could lead to overfitting

File: notebooks/06a Modelling - classification.ipynb [hunk 8]
```diff
@@ -0,0 +1,865 @@
...
+   "outputs": [],
+   "source": [
+    "model_name = \"randomforest\"\n",
+    "\n",
+    "# define gridsearch parameters\n",
+    "gsc = GridSearchCV(\n",
+    "    estimator=RandomForestClassifier(class_weight=weights.to_dict()),\n",
+    "    param_grid={\n",
+    "        \"n_estimators\": [50, 100, 200],  # default 100\n",
+    "        \"max_depth\": [5, 10, None],\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:417]: Having no max depth here could lead to overfitting as with Decision Tree (perhaps less of an issue here due to nature of Random Forest)

File: notebooks/06a Modelling - classification.ipynb [hunk 9]
```diff
@@ -0,0 +1,865 @@
...
+    "coef.set_index(0, inplace=True)\n",
+    "# Plot interactive plot\n",
+    "# Hover over a feature for full feature name\n",
+    "fig = px.bar(coef, x=coef.index, y=\"coef\")\n",
+    "fig.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:824]: Will there be binary classification added? Feels like could be potential for that to perform better

File: notebooks/06a Modelling - classification.ipynb [hunk 10]
```diff
@@ -0,0 +1,865 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "for model in models:\n",
+    "    preds = models[model][\"model\"].predict(X_test)\n",
+    "    probs = models[model][\"model\"].predict_proba(X_test)\n",
+    "    # calculate performance\n",
+    "    balanced_accuracy = balanced_accuracy_score(y_test, preds)\n",
+    "    f1_score_weighted = f1_score(y_test, preds, average=\"weighted\")\n",
+    "    auc = roc_auc_score(\n",
```
Comment (matthewcooper19) [inline notebooks/06a Modelling - classification.ipynb:638]: Could also be worth looking at how precision/recall/f1 score vary by class, to see if this classifier is particularly strong/weak in specific classes

File: notebooks/06a Modelling - classification.ipynb (hunk 9 above)
Comment (amadeuspzs) [reply notebooks/06a Modelling - classification.ipynb:824]: Great suggestion, could do:

Binary classification Long Stayer (21+ days) or not
->
Regression for Short Stay
OR
Regression for Long Stay

Could we do this for IS_MAJOR?

File: notebooks/06a Modelling - classification.ipynb (hunk 10 above)
Comment (amadeuspzs) [reply notebooks/06a Modelling - classification.ipynb:638]: Am adding actual vs predicted plots to all models for now, with note under "Extensions" to add PR curves per class in notebook

File: notebooks/06a Modelling - classification.ipynb (hunk 5 above)
Comment (amadeuspzs) [reply notebooks/06a Modelling - classification.ipynb:302]: Will add in with elasticnet

File: notebooks/06 Modelling - classification.ipynb [hunk 11]
```diff
@@ -0,0 +1,976 @@
...
+    "        model_y_test = y_test\n",
+    "    elif model == \"logreg\":\n",
+    "        model_X_test = X_test_scaled\n",
+    "        model_y_test = y_test\n",
+    "    else:\n",
+    "        model_X_test = X_test\n",
+    "        model_y_test = y_test\n",
+    "\n",
+    "    # perform inference\n",
+    "    preds = models[model][\"model\"].predict(model_X_test)\n",
```
[the remark this answers is not in the document]
Comment (matthewcooper19) [reply notebooks/06 Modelling - classification.ipynb:737]: Sklearn version mismatch, upgrading to latest fixed issue

**Review by matthewcooper19 — APPROVED** [review]
